katana/native/file_manager/file_manager_utils/ftp_utils.py

```diff
@@ -0,0 +1,153 @@
+        # storlines prints new line character at the end of each line in text file
+        # but it does not transfer word files.
+    except:
+        return name + ': Could not transfer the file.'
+    fp.close()
+    return name + ' :Success'
+
+
+def isDir(ftp, path, name, files_path, files_name, all_files_path):
+    """
+    If the selected item is a directory, then this is a recursive function that FTPs all the files selected within
+    the directory. If the item selected is a directory, then this function is called on that directory.
+
+    The Directory structure is recreated in the target machine.
+
+    This also checks if there are any broken tree files. These are files which are not associated to a parent but has
+    some ancestor selected.
+
+    :param ftp: FTP object
+    :param path: Full path of the directory
+    :param name: Name of the directory
+    :param files_path: Full path of the items selected within the directory
+    :param files_name: Names of the items selected within the directory
+    :param all_files_path: all the items in the directory
+    :return: Status or error
+    """
+    result = []
+    parent_dir = ftp.pwd()
+    if parent_dir == '/':
+        new_folder = parent_dir + name
+    else:
+        new_folder = parent_dir + '/' + name
+    try:
+        try:
+            ftp.mkd(new_folder)
+            ftp.cwd(new_folder)
+        except:
+            print('error in creating new directory in target machine')
+            result.append(new_folder + ': Such a Folder already exists')
+            return result
+        files_list = os.listdir(path)
+        try:
+            for item in files_list:
+                # Need to check if the item is in the files_path
+                if path == '/':
+                    item_path = path + item
+                else:
+                    item_path = path + '/' + item
+                if item_path in all_files_path:
+                    try:
+                        if os.path.isdir(item_path):
+                            result.append(isDir(ftp, item_path, item, files_path, files_name, all_files_path))
+                        elif os.path.isfile(item_path):
+                            result.append(transfer_file(ftp, item_path, item))
+                    except:
+                        print('Error in recursive part')
+                # To check if there is any broken files selected
+                elif os.path.isdir(item_path):
+                    broken_files = os.listdir(item_path)
+                    for broken_item in broken_files:
+                        broken_item_path = item_path + '/' + broken_item
+                        if broken_item_path in all_files_path:
+                            print('broken tree file: ' + broken_item_path)
+                            result.append(broken_item_path + ' :broken tree file')
+            ftp.cwd(parent_dir)
+        except:
+            print('Error in for loop')
+    except:
```

Comment:
Please catch specific and generic exceptions and print them out wherever applicable.

Reply:
Done